### DISCUSSION TITLE
Reproducibility of the PDFs

### DISCUSSION THREAD

**Discussion by SrEstegosaurio** [typst/typst#1439] [Q&A] [answered]
As I was unable to find any references to this anywhere I'm posting this question: **Does typst support creating reproducible PDFs?**

This, while not support by default, is possible to do in LaTeX with tooling like nix _—specially flakes—_ and would be a really nice to have feature.

Reference: https://flyx.org/nix-flakes-latex/

**Comment by sitandr** [comment, marked as answer]
As far as I know, Typst aims to be as reproducible as possible: all the functions and methods are "pure" and are guaranteed to produce the same result. That is the behaviour that is important for incremental compilation Typst uses. That was a major concern when, e.g., considering adding pseudorandom generator — it should work only with given seed, no random behaviour "out of nowhere". You can try compiling the same typ file and will get _exactly the same_ pdf, no random ids seem to be added (unlike TeX).

Obviously, there are functions that **depend** on the state of the system they are compiled, like `read` and `today` (coming in next release). Making the files constant is quite easy, but there seems no way to manully set the date for now from compiler (of course, you can set concrete date inside the file itself). Maybe when that date feature will be released that proposal could be considered. Probably, there are some external tools that will match your needs to set the "current" date for compilation.

Finally, there is no package manager for now, but in discussions of its future behaviour there are plans to make it possible creating "typst project" with special files with packages listed and corresponding .lock files for reproducible builds (at the current time they are just discussions).

For now, if you use _same Typst version_ and don't use `read` functions that go to some distant folder on your system, you should get a completly reproducible PDF. However, that is not kind of "official" requirement and it isn't tested in CI (testing uses images), so I suppose it can't be guaranteed to be always that way on all the systems. Probably, you could open an issue if you find such behaiour (though that probably wouldn't be a priority for now).

**Reply by laurmaedje** [reply]
> but there seems no way to manully set the date for now
- `datetime(year: 2023, month: 6, day: 9)` will construct a concrete date
- `datetime.today()` will give you the current date

Otherwise, the answer is spot on!

**Reply by sitandr** [reply]
By the "manual set" I meant kind of "date variable" for compiler to make reproducible _any document_ (so you don't need to go inside), as it is done in the reference link. I'm not sure it is worth of it, though. Probably there are some external tools that will handle most of date cases.

UPD: Fixed my answer to make it more clear, thanks for the edit!

**Comment by u-fischer** [comment]
> no random ids seem to be added (unlike TeX).

They are not random. The creation time of a PDF is an important metadata and so all TeX engines insert if by default. If you recompile the time changes and the ID follows.  That typst doesn't add the date/time as metadata to normal documents and so doesn't have to adjust the ID is in my eyes not a plus but a deficiency. 

When a reproducible build is needed, e.g. for regression tests, there are various commands to do it in TeX. I normally use something equivalent to  `\DocumentMetadata{}\UseName{pdfmeta_set_regression_data:}` at the begin of the document which forces a specific date.

**Reply by sitandr** [reply]
Well, I like that default typst behaviour for now. E.g., it is handy if you are using `git`: pdf files will change only if inner contents are changed, so it is easier to handle commits. Probably, there should be compiler/set options for that, but I don't really know the way that metadata is useful.

**Reply by u-fischer** [reply]
You would be probably quite unhappy if your fotos didn't have metadata saying when you made them -- the file date is not enough as it can easily change when you clone or save.  I seldon save PDF's in git, after all I normally have the source; but even in the cases where the PDF is in git I wouldn't remove such metadata only to make the workflow smoother.

**Reply by SrEstegosaurio** [reply]
> They are not random. The creation time of a PDF is an important metadata and so all TeX engines insert if by default. If you recompile the time changes and the ID follows. That typst doesn't add the date/time as metadata to normal documents and so doesn't have to adjust the ID is in my eyes not a plus but a deficiency.

Options like that one should be possible to tweak, since I don't have any doubts regarding the utility of those options for some people, but having the liberty to modify them is a plus.

For me the current behaviour is quite nice. :)


Also, you need to go out of you way to create fully reproducible PDFs with LaTeX, see the article linked above.

**Reply by u-fischer** [reply]
Sure one should be able to tweak such options, and this is possible in TeX. And no, you don't need to go "out of your way" with pdflatex and lualatex, you can set everything in the document. BTW: in the article they completly remove with pdflatex and lualatex the ID entry. This is not allowed with PDF 2.0 and is not recommended for other versions: 

> Although this entry is optional prior to PDF 2.0, its absence can prevent the PDF file from functioning in some workflows that depend on PDF files being uniquely identified.

**Reply by SrEstegosaurio** [reply]
Oh, my bad then. I'll look more into it.

Thanks for pointing out my mistakes, I'll try to fix it in order to stick to best practices.

**Reply by mmlb** [reply]
Many software tools embed the build date into binaries by default, making reproducible builds impossible. Fortunately there's been lots of work done as part of https://reproducible-builds.org to combat that. One way that I think applies to this conversation is for typst to embed the builddate by default *but* support reading the date from the `SOURCE_DATE_EPOCH` environment variable. Some build tools/scripts will set the envvar to the git commit timestamp automatically, it would be preferable for typst to follow that name instead of going off with their own.

**Reply by laurmaedje** [reply]
Typst already supports `SOURCE_DATE_EPOCH`.

Edit: Saw that you found out and responded in another comment.

**Comment by SrEstegosaurio** [comment]
Thank you all for your replies, this has been quite fast. :D

I'll start working on a new flake to generate my t¡Typst documents right now. Also, my maths teacher is also playing around with Typst now.

**Comment by EduBic** [comment]
If we intend reproducibility as the output bytes are the same between one compilation to another with the same typst input then the PDF is not reproducible. The new file has metadata changed.
![typst_meta](https://github.com/typst/typst/assets/13170564/a8a598cb-4fab-48f9-bdc1-e07f57451b98)

Is it possible to avoid this? 
I have other applications that relies on the hash of the file to understand if a pdf is changed or not.

**Reply by laurmaedje** [reply]
You can `#set document(date: none)` or `#set document(date: datetime(..))` to disable the automatic timestamp creation.

**Reply by EduBic** [reply]
Many thanks, that's exacly what I needed

**Reply by mmlb** [reply]
Also it looks like you can set `SOURCE_DATE_EPOCH` environment variable and that will be picked up automatically: https://github.com/typst/typst/blob/23746ee18901e08852306f35639298ad234d3481/crates/typst-cli/src/world.rs#L404

**Reply by EduBic** [reply]
I'm back in the discussion since I noticed that the compilation output differ between W10 and W11 OS. 
What cause this? Fonts?
What deps has typst compiler?
